server/server.go

```diff
@@ -86,22 +85,23 @@ func NewShipperServer(cfg Config, publisher Publisher) (ShipperServer, error) {
 	return &s, nil
 }
 
-// GetAcceptedIndex returns the accepted index
-func (serv *shipperServer) GetAcceptedIndex() uint64 {
-	return uint64(serv.publisher.AcceptedIndex())
-}
-
-// GetPersistedIndex returns the persisted index
-func (serv *shipperServer) GetPersistedIndex() uint64 {
-	return uint64(serv.publisher.PersistedIndex())
+// getPersistedIndex returns the persisted index, or 0 if there is an error.
+// (An error can only happen if the queue is closed, in which case a placeholder
+// of 0 is often sufficient. Callers that need to explicitly recognize an error
+// state should use publisher.PersistedIndex directly.)
+func (serv *shipperServer) getPersistedIndex() uint64 {
+	index, err := serv.publisher.PersistedIndex()
+	if err != nil {
+		// An error means the queue has been closed, so just return a placeholder.
+		return 0
+	}
+	return uint64(index)
 }
 
 // PublishEvents is the server implementation of the gRPC PublishEvents call.
 func (serv *shipperServer) PublishEvents(ctx context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
 	resp := &messages.PublishReply{
-		Uuid:           serv.uuid,
-		AcceptedIndex:  serv.GetAcceptedIndex(),
-		PersistedIndex: serv.GetPersistedIndex(),
```

Comment:
Do we still need to return `PersistedIndex` if `len(req.Events) == 0`? It was returned before but not anymore after this change.

Reply:
Hmm. Honestly I'd prefer `len(req.Events) == 0` to be a `InvalidArgument` error, which would resolve this as a side effect, any objections?

Reply:
That sounds reasonable to me.

Thinking about it I'm not sure how valuable returning `PersistentIndex` from `PublishEvents` is, it might be worth just removing that entirely (but that can be a separate discussion/PR). The API should ideally have one way to do things so we don't end up having to support multiple use cases.

Reply:
I think I left it in this call with the idea that this way many clients will never need to call `PersistedIndex` (except maybe on shutdown). But yes, we can figure out which pattern makes more sense.

Reply:
I think we should return `PersistedIndex` for `len(req.Events) > 0` here. It's quite convenient for the client – it can determine whether it needs to call the `PersistedIndex` endpoint or assume the batch persisted right after calling the `Publish` endpoint.

If I understood correctly the in-memory queue acknowledges all the events immediately without a delay, so there is no need to call the separate `PersistedIndex` endpoint for this queue implementation and the value in the response can indicate that.

Reply:
> If I understood correctly the in-memory queue acknowledges all the events immediately without a delay, so there is no need to call the separate PersistedIndex endpoint for this queue implementation and the value in the response can indicate that.

Keep in mind that shipper clients have no way to know which queue type they are publishing to, so I'm not sure how they would know they can trust the returned `PersistedIndex` from any individual `Publish` call. It seems simpler to have one single way to obtain `PersistendIndex` that works in all cases, or if we have two ways they should behave exactly the same.

Reply:
Accepted and persisted will ~never match. The memory queue is "persisted" when it has been fully acknowledged by the output, which will probably never happen during the RPC call itself. The disk queue only needs to write to the local disk, but that's still too much to finish in a non-blocking sequence. `PersistedIndex` isn't included for the current batch, but for previous batches. This is for clients that have a publish loop, so they can update their running value of `PersistedIndex` as a side effect of event publication, but it's only useful for recurring calls, it won't say anything about the events that were just pubished.

Reply:
E.g. I publish 100 events and get acceptedIndex 100, persistedIndex 0. I then publish 100 more events and get acceptedIndex 200, persistedIndex 100, since some were persisted in between publish calls.

Reply:
Sounds reasonable to me, let's remove this field from the response and rely on just the `PersistedIndex` endpoint everywhere.

Reply:
I made the API change here https://github.com/elastic/elastic-agent-shipper-client/pull/14